src/main/java/org/jenkinsci/plugins/ccxml/CCXMLAction.java

```diff
@@ -51,14 +60,44 @@ public String getDisplayName() {
 
     @Override
     public String getUrlName() {
-        return "cc.xml2";
+        return URL_NAME;
     }
 
     public View getView() {
         return this.view;
     }
 
     /**
+     * @return A map containing the items in the view object. If the request
+     * contains a query parameter named "recursive", then folders in the view
+     * are traversed recursively and all items in those folders are returned
+     * as well. The map is keyed by the folder the item is in, with top-level
+     * items having an empty key.
+     */
+    @Restricted(NoExternalUse.class)
+    public Collection<TopLevelItem> getItems() {
+        String recursive = Stapler.getCurrentRequest().getParameter("recursive");
+        if (recursive == null) {
+            return view.getItems();
+        } else {
+            return Collections.unmodifiableCollection(getItemsRecursive(view.getItems()));
+        }
+    }
+
+    private Collection<TopLevelItem> getItemsRecursive(Collection<TopLevelItem> items) {
+        List<TopLevelItem> result = new ArrayList<>();
+        for (TopLevelItem i : items) {
+            if (i instanceof ItemGroup) {
+                ItemGroup g = (ItemGroup) i;
+                result.addAll(getItemsRecursive(g.getItems()));
```

Comment:
passing target array as an argument would make the method much more efficient from the memory allocation perspective. Now you allocate a new array list for every Folder/ItemGroup just to move them to another array on the upper level. And so on

Reply:
Yes that's a very good point, I'll fix it. Thanks!